### PR TITLE
Support 'non-contiguous with holes' NJTs for contiguous clone()

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131945
* __->__ #132776
* #131937
* #131704
* #131898

It's possible to construct an NJT with "holes" by specifying both `offsets` and `lengths` metadata. When `nt.clone(memory_format=torch.contiguous_format)` is called on such an NJT, the result should be an NJT without holes.

This PR fixes this in simplistic way using `unbind()`, which isn't really supported in `torch.compile`. The longer term solution involves writing a proper kernel to support this.

NB: Another limitation is that the returned NJT does not have the same ragged structure as the input. While we could manually hack the nested int registry (or update the union find when that lands), this is the first instance where a NJT with holes and an NJT without holes could have the same ragged structure, and getting those to play nicely together requires some fairly involved updates. For now, this PR punts on these updates until we can clean this up.

### CODE REVIEW

File: torch/testing/_internal/opinfo/definitions/nested.py [hunk 1]
```diff
@@ -205,6 +205,27 @@ def _f(op, sample):
...
+        values = torch.randn(
+            10, 5, device=device, dtype=dtype, requires_grad=requires_grad
+        )
+        offsets = torch.tensor([0, 2, 4, 10], device=device, dtype=torch.int64)
+        lengths = torch.tensor([2, 1, 3], device=device, dtype=torch.int64)
+        njt = torch.nested.nested_tensor_from_jagged(
+            values, offsets=offsets, lengths=lengths
+        )
+
+        yield SampleInput(njt, kwargs={"memory_format": memory_format})
```
Comment (soulitzer) [inline torch/testing/_internal/opinfo/definitions/nested.py:226]: Just to double check, would this test fail even if we didn't properly handle the contiguous_format via unbinding?
Maybe I'm missing something, but it seems like when we call `assertEqualIgnoringNestedInts`, it seems like we would end up unbinding anyway.

**Review by ani300 — APPROVED** [review]
lgtm

File: torch/nested/_internal/ops.py [hunk 2]
```diff
@@ -421,6 +421,37 @@ def is_contiguous_general(func, *args, **kwargs):
...
+        if new_kwargs["memory_format"] == torch.contiguous_format:
+            # need to copy to remove "holes" non-contiguity / lengths metadata
+            # TODO: write a kernel for this
+            from .nested_tensor import jagged_from_list
+
+            # TODO: We probably want the output to have the same ragged structure / nested int.
+            assert (
+                inp._ragged_idx == 1
+            ), "NJT with ragged_idx != 1 not supported for contiguous clone"
+            contig, _ = jagged_from_list(inp.unbind(), offsets=None)
```
Comment (ani300) [inline torch/nested/_internal/ops.py:446]: this would be as "simple" as having the dim that used to be referenced by the lengths be used by the offsets here, right?

Comment (ani300) [reply torch/nested/_internal/ops.py:446]: (referring to the TODO)

Comment (jbschlosser) [reply torch/nested/_internal/ops.py:446]: almost (I tried this briefly), I think there's more work required throughout to actually support ops with inputs that have the same nested int but different contiguity. We have to be smarter about how we operate on the `values` buffer when the lengths metadata is present.

Comment (ani300) [reply torch/nested/_internal/ops.py:446]: I see... I guess it can wait for a further PR then

File: torch/testing/_internal/opinfo/definitions/nested.py (hunk 1 above)
Comment (jbschlosser) [reply torch/testing/_internal/opinfo/definitions/nested.py:226]: yep, I verified that the test fails before my `contiguous_format` handling logic. The reason why is that before this PR, the lengths metadata wouldn't be propagated and so unbinding would just give you the wrong results (not taking into account lengths).

Comment (soulitzer) [reply torch/testing/_internal/opinfo/definitions/nested.py:226]: I see, thanks

**Review by soulitzer — APPROVED** [review]
Sounds good!
